### DISCUSSION TITLE
bug: Prefetching with SSG helpers still makes the query fetch on the client since beta.14 (`v10`)

### DISCUSSION THREAD

**Discussion by bennettdams** [trpc/trpc#2925] [Q&A] [answered]
### Provide environment information

```
  System:
    OS: Linux 5.10 Ubuntu 20.04.5 LTS (Focal Fossa)
    Container: Yes
    Shell: 5.8 - /usr/bin/zsh
  Binaries:
    Node: 16.17.0 - /usr/bin/node
    npm: 8.15.0 - /usr/bin/npm
  npmPackages:
    @tanstack/react-query: ^4.9.0 => 4.9.0 
    @trpc/client: ^10.0.0-proxy-beta.13 => 10.0.0-proxy-beta.13 
    @trpc/next: ^10.0.0-proxy-beta.13 => 10.0.0-proxy-beta.13 
    @trpc/react: ^10.0.0-proxy-beta.13 => 10.0.0-proxy-beta.13 
    @trpc/server: ^10.0.0-proxy-beta.13 => 10.0.0-proxy-beta.13 
    next: 12.3.1 => 12.3.1 
    react: ^17.0.2 => 17.0.2 
    typescript: 4.8.4 => 4.8.4
```

### Describe the bug

Prefetching via SSG Helpers is broken since `10.0.0-proxy-beta.14`.
No matter if you prefetch on the server or not, the query will be refetched on the client.

Some particular change that maybe is responsible: https://github.com/trpc/trpc/commit/1d187b11f83dade3a0e5643a66b3b4fe5974070c

This change happened for beta.14, and the prefetching was still working in beta.13.

### Link to reproduction

https://github.com/bennettdams/trpc-10-unmounted-bug

## Via reproduction repo :

(the repo was created for another issue, but works for this issue as well)

Steps:

1. `npm i` (installs beta.13)
2. `npm run dev`
3. Go to <http://localhost:3000/user/123>
4. See in the devtools that the prefetched query is **not** fetched on the client.
5. `npm install-14` (to install beta.14)
6. Restart, `npm run dev`
3. Go to <http://localhost:3000/user/123>
4. See in the devtools that the prefetched query **is** fetched on the client.

![image](https://user-images.githubusercontent.com/29319414/194181230-cc35211f-6e40-4cce-b167-6cccd2c4346a.png)

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment]
This depends on your query client config in react-query and isn't related to tRPC as such.

**Reply by bennettdams** [reply]
Could you be more specific?
Mabye you were looking at an old version of my reproduction, but I [disabled refetch on mount/focus](https://github.com/bennettdams/trpc-10-unmounted-bug/commit/f37e538bf5feccb8b792c480f4a950a7d803e38f) half an hour ago.

Also, this was working perfectly fine in beta.13, but stopped working in beta.14, so what change is expected in the React Query config?

**Reply by KATT** [reply]
You probably wanna set a `staleTime` https://tanstack.com/query/v4/docs/guides/initial-query-data#staletime-and-initialdataupdatedat

**Reply by bennettdams** [reply]
I'm not sure about that 🤔 If the client is created for each `getStaticProps` invocation with a [revalidation time of 5 minutes](https://github.com/bennettdams/trpc-10-unmounted-bug/blob/master/src/pages/user/%5BuserId%5D.tsx#L37) and the cache is populated there, it shouldn't change until next revalidation - or in other words, I think the prefilled cache should still be valid and not be refetched, because not every route visit will execute `getStaticProps` and therefore the same client is used. I even wrote about it here: https://twitter.com/BennettDams/status/1459184917192093700 and the [docs also mention it in the caveats](https://tanstack.com/query/v4/docs/guides/ssr#staleness-is-measured-from-when-the-query-was-fetched-on-the-server).

..but this was for vanilla React Query with manual hydration where the client was created on each invocation. I'm not sure about the lifecycle of tRPC's query client on the server. It at least looks like a new query client is created for each invocation of [createSSGHelpers](https://github.com/trpc/trpc/blob/3b957c1bd1b9dc5fe699f00d885f81af43574928/packages/react/src/ssg/ssg.ts#L37).

Nevertheless, I [added a long stale time for the query at the reproduction repo](https://github.com/bennettdams/trpc-10-unmounted-bug/commit/c076bbbdb39ffddfa7ba845877fbbba85477b4a9), and the problem is the same.
And again, this was working for beta.13, but doesn't work anymore with beta.14 😅

**Reply by bennettdams** [reply]
@KATT  I tested this again [with beta.17](https://github.com/bennettdams/trpc-10-unmounted-bug/commit/abd90c416c66ccea56ee9168c43ca73b929ea031) and the described problem remains.

Could you let me know why you don't consider this a bug? I think it is and the consequences are that one cannot use prefetching/prepopulating the cache on the server anymore, which is not intended, right?

**Reply by bennettdams** [reply]
Update: [I added a `getServerSideProps` route](https://github.com/bennettdams/trpc-10-unmounted-bug/commit/209588896d19557a484b3e88438c7221d51c1dc7) and it is also not working there.

**Comment by freeatnet** [comment, marked as answer]
@bennettdams I see the same issue.

Here's the simplest repro I found:
1. Check out [trpc/examples-next-prisma-todomvc](https://github.com/trpc/examples-next-prisma-todomvc) (which uses `^10.0.0-proxy-beta.17`).
2. Log `allTasks.status` right after [`trpc.todo.all.useQuery(…)`](https://github.com/trpc/examples-next-prisma-todomvc/blob/next/src/pages/%5Bfilter%5D.tsx#L174).
3. Run the server.
4. Navigate to `http://localhost:3000` (or better yet, curl it).

Expected: the status is reported as `success` and the query-provided data is used to generate the page contents.
Observed: the status is reported as `loading` and the query-provided data is missing from the page. If I downgrade to `@trpc/*@10.0.0-proxy-beta.13`, expected behaviour is restored.

After a bit of debugging, it looks like `getArrayQueryKey`, which is [used to normalize the hooks' `pathAndInput`](https://github.com/trpc/trpc/commit/1d187b11f83dade3a0e5643a66b3b4fe5974070c), should also be applied to `pathAndInput` in the [SSG helpers.](https://github.com/trpc/trpc/blob/d6d4c8fc43003fcf3f3876d676d66a99c84633e7/packages/react/src/ssg/ssg.ts#L81)

@KATT I opened a PR with the fix, https://github.com/trpc/trpc/pull/2976 — please let me know if you have any suggestions. I'd also like to add an integration test to make sure SSG prefetch state is readable by React hooks, would that be a sensible thing to do?

**Comment by freeatnet** [comment]
@bennettdams this issue has been fixed in [v10.0.0-proxy-beta.18](https://github.com/trpc/trpc/releases/tag/v10.0.0-proxy-beta.18). Let me know if it works for you.

**Reply by bennettdams** [reply]
Works great, thanks a lot.
